cmds/core/msr/msr.go

```diff
@@ -1,88 +1,210 @@
+	}
+}
+
+func and(f forth.Forth) {
+	v := f.Pop().(uint64)
+	m := f.Pop().([]uint64)
+	forth.Debug("and: %v(%T) %v(%T)", m, m, v, v)
+	for i := range m {
+		m[i] &= v
+	}
+	f.Push(m)
+}
 
-	case "w":
-		// Sadly, we don't get an error on write if the values
-		// don't match.  Reading it back to check it is
-		// not always going to work. There are many write-only
-		// MSRs. If there are no errors there is still no
-		// guarantee that it worked, or if we read it we would
-		// see what we wrote, or that vmware did not do
-		// something stupid, or the x86 did not do something really
-		// stupid, or the particular implementation of the x86
-		// that we are on did not do something really stupid.
-		// Why is it this way? Because vendors hide proprietary
-		// information in hidden MSRs, or in hidden fields in MSRs.
-		// Checking is just not an option. There, feel better now?
-		if len(a) < 4 {
-			log.Fatal(usage)
+func or(f forth.Forth) {
+	v := f.Pop().(uint64)
+	m := f.Pop().([]uint64)
+	forth.Debug("or: %v(%T) %v(%T)", m, m, v, v)
+	for i := range m {
+		m[i] |= v
+	}
+	f.Push(m)
+}
+
+func main() {
+	flag.Parse()
+	if *debug {
+		forth.Debug = log.Printf
+	}
+
+	// TODO: probably can do this by just having two passes, and write
+	// in the first pass is a no op. Which will fail to catch the problem
+	// of read-only and write-only MSRs but there's only so much you can do.
+	//
+	// To avoid the command list from being partially executed when the
+	// args fail to parse, queue them up and run all at once at the end.
+	//queue := []func(){} etc. etc.
+	f := forth.New()
+	for _, o := range ops {
+		forth.Putop(o.name, o.op)
+	}
+	for _, w := range words {
+		forth.NewWord(f, w.name, w.w[0], w.w[1:]...)
+	}
+	a := flag.Args()
+	switch {
+	case len(a) == 3 && a[0] == "r":
+		// Because the first arg is a glob and may have things like * in it (* being the
+		// most common) gratuitiously add a Forth ' before it (i.e. quote it).
+		if err := forth.EvalString(f, fmt.Sprintf("'%s msr %s reg rd", a[1], a[2])); err != nil {
+			log.Fatal(err)
 		}
-		v, err := strconv.ParseUint(a[3], 0, 64)
-		if err != nil {
-			log.Fatalf("%v: %v", a, err)
+	case len(a) == 4 && a[0] == "w":
+		// Because the first arg is a glob and may have things like * in it (* being the
+		// most common) gratuitiously add a Forth ' before it (i.e. quote it).
+		if err := forth.EvalString(f, fmt.Sprintf("'%s msr %s reg rd", a[1], a[2])); err != nil {
+			log.Fatal(err)
 		}
-		errs := wrmsr(m, uint32(reg), v)
-		for i, e := range errs {
-			if e != nil {
-				fmt.Printf("%v: %v\n", m[i], e)
+	default:
+		for _, a := range flag.Args() {
+			if err := forth.EvalString(f, a); err != nil {
+				log.Fatal(err)
 			}
+			forth.Debug("%vOK\n", f.Stack())
 		}
-
-	default:
-		log.Fatalf(usage)
+	}
+	// special case: if the length of stack is 1, just print out stack[0].
+	s := f.Stack()
+	if len(s) == 1 {
+		fmt.Printf("%v\n", s[0])
+	} else {
+		fmt.Printf("%v\n", f.Stack())
```

Comment:
Maybe use `s` here instead of `f.Stack()` ?

Reply:
done